### ISSUE TITLE
[SVGKImage imageNamed:]  [CFCharacterSet invertedSet]: message sent to deallocated instance

### ISSUE THREAD

**Issue by cohen72** [SVGKit/SVGKit#85]
I switched to use cocoapods to install SVGKit.  Since doing so, I'm crashing when using  

UIImage\* newImage = [SVGKImage imageNamed:@"mySVGfile"].UIImage;

Error I receive is:

[CFCharacterSet invertedSet]: message sent to deallocated instance

Any ideas as to why this is happening suddenly?

**Comment by MaddTheSane** [comment]
Does it crash without the UIImage call?

**Comment by adamgit** [comment]
1. That might be ONLY a problem with CocoaPods - currently, I don't think anyone's using CP, so no-one's testing it :(.
2. It sounds like CocoaPods is pointing to a random commit that it shoudn't be - that bug was fixed within a couple of days of being committed, and ... that was 2 months ago.

What commit is CocoaPods using?

**Comment by adamgit** [comment]
OK, it looks like CocoaPods might be using git-tags instead of git-branches. That makes sense, but it only works for projects that have hard version numbers.

It seems CP has to be manually changed every time there's any patch?. Unless there's an easy way to force CocoaPods to read a branch instead of a tag ... I'm tempted to say "CocoaPods is unsupported, it's far too much effort to keep updating it all the time" :(.

**Comment by LeChatNoir69** [comment]
Hi there,

I've installed SVGKit 1.0 with CocoaPod. I've fixed two issues :
=> this one : iPhone 5s has the same density as 5 and 4
=> and this one : Hotfix for CGPathAddRoundedRect compile error on iOS 7.

Now, my SVG is well loaded, well displayed and well animated. But when I try to load it a second time, it crashes with this message : 
[CFCharacterSet invertedSet]: message sent to deallocated instance

It's happening in parseData method of SVGPathElement.

Any idea ?

**Comment by LeChatNoir69** [comment]
Looking deeply, I come to this point : 
- (CGPoint) readMoveto:(NSScanner*)scanner path:(CGMutablePathRef)path relativeTo:(CGPoint)origin isRelative:(BOOL) isRelative

In SVGKPointsAndPathsParser.m.
Crash happens when this method is called :     [SVGKPointsAndPathsParser readWhitespace:scanner];

But it makes no sense cause scanner seems ok...
It seems there is another object released at this point but can't figure out what :-(

**Comment by adamgit** [comment]
As per previous comments: CocoaPods are not supported. You're using old
code, and re-fixing bugs that were fixed months ago :(.

If you go to the front page of the project, you will always get the current
development branch, with latest fixes. Currently 1.2.x.

Each of the items you describe sounds like things that were fixed in 1.1.0.

I am tempted to delete the CocoaPods info, if I can find out how. The
person who added it isn't around to maintain it, and it's not helping :(.

**Comment by LeChatNoir69** [comment]
Yep, you're right. I've tried with 1.x and theses bugs have disappeared.

I think we should tag the 1.1.0 branch and update the podspec. I'm not an expert but I'd like to try. What do you think ?

**Comment by adamgit** [comment]
Go for it :).

**Comment by LeChatNoir69** [comment]
Ok. I've tried with the 1.1.0 branch but when I tried to test in local, I've got compilation errors. The script post compilation seems to have change ?! I've got this 
.../SVGKit-1.1.0/Source/Vendor/Lumberjack/DDLog.h:268:1: Duplicate interface definition for class 'DDLog'

**Comment by adamgit** [comment]
If you click on the error in the list, Xcode will tell you exactly where
the two definitions are.

Usually its obvious which one is wrong.

**Comment by LeChatNoir69** [comment]
Yes, usually. But here, it's a script problem (the shell run after compilation ?). Error are "Shell script invocation error duplicate interface definition for class "DDLog". (See attachment). Where is the shell script saved ? In the xcodeProj file ?

![capture decran 2013-11-11 a 22 59 26](https://f.cloud.github.com/assets/1202490/1517028/e8b9ca6e-4b1c-11e3-864e-c9e77f662691.png)

**Comment by LeChatNoir69** [comment]
The point is that on this morning, it was ok. Tonight, it's not. I'm loosing my words...

**Comment by adamgit** [comment]
I don't know how CocoaPods works. It sounds like it might be interfering
with Xcode's build system.

Xcode's build system is quite simple. If you look at the "project" window,
and go to the Build Phases tab, the final phase is a Run Script. That
script is heavily documented, but briefly:
- builds whichever version you did NOT build (sim or device)
- creates a Universal binary including ALL versions
- outputs the binary + all header files to a new folder (*-universal)

...this is a workaround to a bug Apple created in Xcode 4.2, where they
deliberately broke Universal builds on iOS.

**Comment by LeChatNoir69** [comment]
No, no cocoaPod here. I've just clone the 1.1.0 branch on my desktop and tried to compile. Don't understand theses errors. It seems to be a problem on my XCode...

**Comment by adamgit** [comment]
DDLog is the new logging framework someone added. It's working fine
normally.

But since its a logging framrework, its normal to #import it into
Prefix.pch, which casues it to be imported into EVERY header in your
project.

Sometimes that causes probelms if you forget, and try ti include it
somewhere else too.

What are the two places Xcode says its being defined?

**Comment by LeChatNoir69** [comment]
Oh god, you're right... I've commented #import "DDLog.h" in all files except .pch and it's fine. (ContextFilterFormatter.h, DDAbstractDatabaseLogger.h, DDASLLogger.h, etc..).

What I don't understand is why is this happening in the 1.1.0 branch. 

This branch didn't change theses past days ? 
So why am I getting theses errors now ?

**Comment by LeChatNoir69** [comment]
In fact, I'd like to use cocoaPod because I can't get SVGKit work for devices using your instructions. It's all right when I'm running my app under simulator but when I'm trying to run on a device, the #import <SVGKit/SVGKit.h> doesn't work anymore.

I've seen it's because you have a "copy header files" phase instead of a simple "Copy files" as it is explained here 
https://developer.apple.com/library/ios/technotes/iOSStaticLibraries/Articles/creating.html#//apple_ref/doc/uid/TP40012554-CH2-SW1

But I can't figure out how to make it work... So I've seen that SVGKit was "CocoaPodised" so I've tried and it works very well under simulator AND device. But unfortunately, it's a very old version...

So if you got a tips to make it work on device with just the static library, I would be happy.

**Comment by adamgit** [comment]
"Copy Headers" is the official way from Apple to copy headers, and is
integrated into Xcode; Copy Files is not. Copy Files is also badly broken -
it doesn't actually copy files! (it does something proprietary that Apple
broke a while back).

Couple of things to check:
- Xcode 4 or 5? Xcode5 has _MAJOR_ bugs right now (e.g. I'm using the
  prerelease of 5.0.2 because 5.0.0 and 5.0.1 are so badly broken they reboot
  my machine every few minutes!). Apple is pushing out updates quickly, which
  is good.
- When you say the lib doesnt work - do you mean "in your own app project"
  or "in the Demo-iOS project" ?
- If it works in Demo-iOS, but not in your app ... go to the Build
  Settings page, and copy paste each row that is BOLD hilighted (Bold means
  "you have changed the default setting"). It might be obvious that one of
  these has somehow got mangled.

... NB: one of the MAJOR bugs in Xcode5 is that it CORRUPTS YOUR BUILD
SETTINGS for static libraries every time you add a new library. This is
very very annoying. I've seen two different ways it corrupts your build
file :(.

**Comment by LeChatNoir69** [comment]
Hmmmm. I disagree with you about the first point : in the doc "How to make
a static library", you can read that (follow the link I copy above) :
"If your library target has a “Copy Headers” build phase, you should delete
it; copy headers build phases do not work correctly with static library
targets when performing the “Archive” action in Xcode"

I'm under XCode 5.0.0.
When I say "it doesn't work", I mean in my application.
I can run it under simulator. And everything is fine.

But when I'm choosing device, there is pb around header files so I can't
build it.

I'll try it again later. Thank you for your patience !

2013/11/12 adamgit notifications@github.com

> "Copy Headers" is the official way from Apple to copy headers, and is
> 
> ## integrated into Xcode; Copy Files is not. Copy Files is also badly broken
> 
> it doesn't actually copy files! (it does something proprietary that Apple
> broke a while back).
> 
> Couple of things to check:
> - Xcode 4 or 5? Xcode5 has _MAJOR_ bugs right now (e.g. I'm using the
>   prerelease of 5.0.2 because 5.0.0 and 5.0.1 are so badly broken they
>   reboot
>   my machine every few minutes!). Apple is pushing out updates quickly,
>   which
>   is good.
> - When you say the lib doesnt work - do you mean "in your own app project"
>   or "in the Demo-iOS project" ?
> - If it works in Demo-iOS, but not in your app ... go to the Build
>   Settings page, and copy paste each row that is BOLD hilighted (Bold means
>   "you have changed the default setting"). It might be obvious that one of
>   these has somehow got mangled.
> 
> ... NB: one of the MAJOR bugs in Xcode5 is that it CORRUPTS YOUR BUILD
> SETTINGS for static libraries every time you add a new library. This is
> very very annoying. I've seen two different ways it corrupts your build
> file :(.
>

**Comment by adamgit** [comment]
Yep, that's ANOTHER bug Apple introduced recently in a version of Xcde4. I
think that was Xcode4.6

There are workarounds which work fine, but last time I checked, "Copy
files" is not a valid workaround. On projects where we've needed to do
this, we've found that wrintg a Run Script that just does "cp from to"
works better :).

Xcode 5.0.0 - definitely upgrade to latest! Tehre are important fixes in
5.0.1

Re: app - go through the build settings, there shoul donly be approx 5 or 6
bolded items, and copy/paste them here, and I'll check tem.

**Comment by LeChatNoir69** [comment]
Ok. I'll keep you update (is this real english expression ? I'm french :))

2013/11/12 adamgit notifications@github.com

> Yep, that's ANOTHER bug Apple introduced recently in a version of Xcde4. I
> think that was Xcode4.6
> 
> There are workarounds which work fine, but last time I checked, "Copy
> files" is not a valid workaround. On projects where we've needed to do
> this, we've found that wrintg a Run Script that just does "cp from to"
> works better :).
> 
> Xcode 5.0.0 - definitely upgrade to latest! Tehre are important fixes in
> 5.0.1
> 
> Re: app - go through the build settings, there shoul donly be approx 5 or
> 6
> bolded items, and copy/paste them here, and I'll check tem.
>

**Comment by LeChatNoir69** [comment]
Ok. This is what I've done :
I've followed the instructions except that I've changed my #import.
Instead of #import <SVGKit/SVGKit.h>, I've input #import "SVGKit.h"

And I've added /usr in my Header Search Path.

Now, if we consider to make SVGKit "Cocoapod aware", there is some things you'll have to change.

For exemple, the log framework is CocoaLumberjack. CocoaLumberjack is under Cocoapod…

So if we want SVGKit to be under cocoa pod, we have to delete the CocoaLumberjack sources from the project and just put CocoaLumberjack as a pod dependencie of SVGKit (this is the right way and avoid duplicate symbols and ease maintenance…).

I think it should be done but it involves a better "adhesion" (or joining ?) of the authors.

**Comment by adamgit** [comment]
Lumberjack is designed to be embedded in projects. It doesnt need to be
separate in CP.

(Someone tried packaging it externally, but that apparently caused problems
- it seems to work better when tightly integrated?)

Does CP work if lunberjak is embedded?

**Comment by adamgit** [comment]
That import change is a huge change. Are you saying you had to do that to
make it work for you?

**Comment by LeChatNoir69** [comment]
Yes, i had to make this change in ordre to make the static library work on device.
When you say it's a huge change, I don't Know if it's ironic (cause it's just few characters :)) or not (cause effectively, it's not the same way to include a framework)

Guillaume

**Comment by LeChatNoir69** [comment]
Lumberjack is already under cocoapod so yes.
For exemple, in a simple project, if I want to use lumberjack, I just have to make a profile with one line :
Lumberjack => "1.0" and then to run pod install and that's it !

Cocoapod is definitely future of library.

Guillaume

**Comment by adamgit** [comment]
I mean: does it work ok with lumberjack NOT coming from cocoappd?

Making it part of CP would _break the library_ for everyone else. I dont
want to do that

**Comment by LeChatNoir69** [comment]
It won't be a "clean" thing to use CocoaPod for SVGKit but let the
Lumberjack code inside SVGKit...
So I'm using the static library now for my project and it will be fine.

Now, for the future of SVGKit, I think it would be a good thing to manage
it with Cocoapod. But we'll have to delete lumberjack source. And you're
not ready for that apparently.

2013/11/13 adamgit notifications@github.com

**Comment by adamgit** [comment]
FYI - when the next release of Xcode5 goes live, I'll review the header
files arrangement, see if there's a cleaner way that avoids the Xcode5 bugs
(probably not, but it's worth a try!)

I won't have time for a few weeks yet, though. So I'll leave this issue
"open", but only because I want to come back and double-check your #import
changes :)

**Comment by LeChatNoir69** [comment]
ok. Thank you guy. For everything :)

2013/11/13 adamgit notifications@github.com

> FYI - when the next release of Xcode5 goes live, I'll review the header
> files arrangement, see if there's a cleaner way that avoids the Xcode5
> bugs
> (probably not, but it's worth a try!)
> 
> I won't have time for a few weeks yet, though. So I'll leave this issue
> "open", but only because I want to come back and double-check your #import
> changes :)
>
